pkg/apis/networking/validation/validation.go

```diff
@@ -597,3 +597,107 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
 	}
 	return false
 }
+
+// ValidateClusterCIDRConfigName validates that the given name can be used as an
+// ClusterCIDRConfig name.
+var ValidateClusterCIDRConfigName = apimachineryvalidation.NameIsDNSLabel
+
+// ValidateClusterCIDRConfig validates a clusterCIDRConfig.
+func ValidateClusterCIDRConfig(ccc *networking.ClusterCIDRConfig) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&ccc.ObjectMeta, false, ValidateClusterCIDRConfigName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateClusterCIDRConfigSpec(&ccc.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateClusterCIDRConfigSpec validates clusterCIDRConfig Spec.
+func ValidateClusterCIDRConfigSpec(spec *networking.ClusterCIDRConfigSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if spec.NodeSelector != nil {
+		allErrs = append(allErrs, apivalidation.ValidateNodeSelector(spec.NodeSelector, fldPath.Child("nodeSelector"))...)
+	}
+
+	maxIPv4PerNodeMaskSize := int32(32)
+	maxIPv6PerNodeMaskSize := int32(128)
+
+	hasIPv4 := spec.IPv4 != nil
+	hasIPv6 := spec.IPv6 != nil
+
+	// Validate if CIDR is configured for at least one IP Family(IPv4/IPv6).
+	if !hasIPv4 && !hasIPv6 {
+		allErrs = append(allErrs, field.Required(fldPath, "one or both of `ipv4` and `ipv6` must be configured"))
+	}
+
+	// Validate configured IPv4 CIDR and PerNodeMaskSize.
+	if hasIPv4 {
+		allErrs = append(allErrs, validateCIDRConfig(spec.IPv4, fldPath, maxIPv4PerNodeMaskSize, "ipv4")...)
+	}
+
+	// Validate configured IPv6 CIDR and PerNodeMaskSize.
+	if hasIPv6 {
+		allErrs = append(allErrs, validateCIDRConfig(spec.IPv6, fldPath, maxIPv6PerNodeMaskSize, "ipv6")...)
+	}
+
+	// Validate PerNodeMaskSize if both IPv4 and IPv6 CIDRs are configured.
+	// IPv4.PerNodeMaskSize and IPv6.PerNodeMaskSize must specify the same number of IP addresses:
+	// 32 - spec.IPv4.PerNodeMaskSize == 128 - spec.IPv6.PerNodeMaskSize
+	// Unequal allocatable IP addresses will lead to wastage of IP addresses.
+	// Also, this being one of the tie-breaker rules for ClusterCIDConfigs, it is
+	// necessary to have equal number of allocatable IP addresses per node.
+	if hasIPv4 && hasIPv6 {
+		if maxIPv4PerNodeMaskSize-spec.IPv4.PerNodeMaskSize != maxIPv6PerNodeMaskSize-spec.IPv6.PerNodeMaskSize {
```

Comment:
I don't know if api machinery validates for negative nums. if it does not this can give true negatives.